### ISSUE TITLE
Grids, Sections & Regions

### ISSUE THREAD

**Issue by benhaynes** [directus/app#773]
#### :rotating_light: MIGRATED FROM REQUESTS.GETDIRECTUS.COM :rotating_light: 

# Grids, Sections & Regions
### :+1: = 1
*Created 9 months ago by @marcobiedermann*
 

-----
### @marcobiedermann – :+1: (9 months ago)
It would be super great it there would be an option to create layouts / grids / sections to nest modules inside. Take a look at Cockpit for reference: https://twitter.com/getcockpit/status/898848619020914688

**Comment by rijkvanzanten** [comment]
This is a duplicate of https://github.com/directus/app/issues/506, https://github.com/directus/app/issues/759, and https://github.com/directus/app/issues/763.

**Comment by rijkvanzanten** [comment]
Groups can be nested as well, so we could make a tab-group that works with nested tab-groups if you know what i mean

```
tabs
  tab
    title
    content
  tab
    color
    map
```

The outer most tab would control the visibility of the inner tabs, which in turn group the actual interfaces

**Comment by benhaynes** [comment]
Beautiful.

So this would be a `Tag Group` interface? And it would be self-aware if it is the parent (tab container) or a sibling (tab)?

What a great/easy solution...

**Comment by rijkvanzanten** [comment]
AFAIK every component can read info on their parent and it's children components. That would make this possible as the parent tab could check for the existence of tab components in it's children and the children can check if the parent is a tab!

**Comment by benhaynes** [comment]
Perfect. So this is just a single interface!

**Comment by benhaynes** [comment]
This is the more robust ticket, so I'm closing the other one. Here is the relevant info from that ticket:

```
1-1-1-1
1-1 - 2
2  -  2
4     -
FUUUUUUUUUUULL (0?)```

**Comment by benhaynes** [comment]
Maybe this is obvious, but we should have this grid system be fully responsive. That means we'll likely want to wrap the layout in half at a certain point: 4->2 and 2->1

**Comment by alrick** [comment]
@rijkvanzanten I see some fields are grouped at the demo (multiple fields on the same line).
How to do this ? can't find anything in the documentation.

**Comment by rijkvanzanten** [comment]
@alrick That's an initial test of this using the `width` column in the DB. While it technically works, we're not too happy with it yet, as it doesn't provide enough flexibility in certain use cases.

You can set the width column in directus_fields to either 1, 2, 3, or 4. It will flex align the items on either 1/4, 2/4, etc

**Comment by alrick** [comment]
@rijkvanzanten Thanks a lot for the tip, that's not too flexible but that's a starter :) I hope you find a nice solution.

**Comment by benhaynes** [comment]
Here are the questions to answer before we can finalize the grid structure and build the managing settings interface:

1. ** Form Width**
    * What is the parent form width? Assuming that most fields will default to full width this should not be too wide or too narrow.
    * Body text fields should have about 50-100 characters per line, with 66 as the "ideal"
    * Maps, Charts, etc should have enough room to feel immersive and complete (subjective)
    * This form width will also determine Column size, so it should be evenly divisible by the desired column width/count without the smallest column with being too narrow.
2. **Column Size Options**
    * Currently we have 4 options (25%, 50%, 75%, 100%), but we could potentially simplify this to 2 options (50% or 100%). I know that 4 columns offers a lot more flexibility, but it means having pretty small columns that might not fit longer labels or notes.
3. **Allow Blank**
    * Do we allow for blank areas or do we simply _float_ all fields to the left?
4. **Fixed/Percentage Widths**
    * Should these be fixed width, percentage-based, or a combination (responsive). If there is any percentage fluidity then _all_ interfaces need to support this. If fixed, what happens when the viewport is smaller than the form width?
5. **Responsive Breaks**
    * Will we allow for fields to break down to the next row/line when their widths don't fit within the viewport? If so, then technically the grid that admins setup would only be present when in larger/desktop sizes. With either 2 or 4 column options I think we always need to split the fields in _half_... meaning that if we have 4 fields across then it would wrap to 2/2 instead of 3/1.
6. **Max Width**
    * Do we allow for a "Full" width width that extends beyond the columns to fill 100% of the remaining viewport? Or do _all_ fields stay within the parent form width?
7. **Row Height**
    * Do we want this to be adjustable? Most interfaces use our standard inputs which all have the same height (40px as of now). Most "full-width" type interfaces (eg: Map, WYSIWYG, etc) will support taller and dynamic heights.
    * Do we need to set heights in the layout, or leave this to interfaces?
8. **Padding**
    * Each column must be equally spaced based on global rules. This means that if we were to have 2 column options (50% and 100%), the full width would equal: `col-width + padding + col-width`. This gets slightly more complex with 4 column options.
9. **Notes**
    * As of now, each field is stacked as: Label, Note (optional), Interface. The problem is that the Note is both optional and variable in height. So if you have two input fields side-by-side and only one has a note, you end up with the fields misaligned (see below). The user could try to adjust Notes to avoid this, but with variable lengths and responsiveness this will be tricky. We need a more fool-proof method — such as adding Notes _below_ the interface, at the bottom of the field.

![screen shot 2018-12-04 at 9 51 09 am](https://user-images.githubusercontent.com/522079/49450035-2dd16a00-f7aa-11e8-8141-f757aa5d4c36.png)


**Is there anything else we need to discuss/answer?**

**Comment by benhaynes** [comment]
I didn't want to muddle the questions with my proposed answers, but here are my thoughts:

1. Form Width — 630px
2. Column Sizes — Only two: 50% and 100%. I think the 25%/75% sizes add flexibility but will make grids needlessly complex and less uniform. KISS.
3. Allow Blank — Yes, but only on the right. This would be easier to save with only two width options, we simply would have something like: `50`, `50-`, and `100`. (also, `100` could be the default and therefore `NULL` could be used to simplify things further)
4. Fixed/Percentage Widths — Fixed Form, Fixed Fields... until viewport forces the form smaller than 300px. (see below)
5. Responsive Breaks — With only two sizes, we simply break the side-by-side fields when they no longer fit beside each other. Once the form has less than 300px width we can make all fields 100% width _or_ force x-scroll in the form (to avoid any % widths).
6. No. I think 630px is enough for almost all use-cases, and interfaces that demand more space can use a "full-screen mode". For example a WYSIWYG editor could be 630px across and 800px tall... with a Distraction Free toggle button to go full screen as needed. Keeping the form at a fixed width leaves us room to show the Info Sidebar.
7. Row Height — No. All single-row interfaces should be the standard height. Other interfaces, such as WYSIWYG, Maps, Charts, and Relational Tables should have individual options (using standardized units!) to adjust height or max-height.
8. Padding — 30px column (horizontal) and 40px row (vertical), per the "v8" design mock standard
9. Notes — Rendered _below_ each interface

The above options actually significantly simplify the way we configure the Form Grid. All we would have to do is change the allowed widths from: `1`, `2`, `3`, and `4` ... to: `50` and `50-`. The rest is accomplished with CSS updates to the App. 😄 

With all of this together you get a very flexible, yet still standardized and fool-proof form layout such as this:

![screen shot 2018-12-04 at 10 27 31 am](https://user-images.githubusercontent.com/522079/49452493-4b550280-f7af-11e8-8006-8432b55cbeac.png)

**Comment by ceruulean** [comment]
I made a prototype "field arranger" that uses the current information structure:
https://streamable.com/8530g

The basis is that it reflects what is in the database, so moving fields changes their sort order, and the side handlebars changes their width to 4 predefined sizes (according to directus global values). There are guidelines to help the user know where fields snap to size.

* The text-input interface has an additional option for 'width', so that reflects in the preview. You can set the width of the box itself, while the interface width is something different.

- Resize works on Firefox but does not show the change until 'dragEnd'
- On Chrome, the left handlebar will stop working sometimes, but the right handlebar always works.
- The 'sort' checkbox is WIP, supposed to be a toggle between "sort" and "group" (dragging in 'sort mode' changes the position, but dragging in 'group mode' will add fields to groups) but I haven't implemented grouping because there's an actual 'group' interface and then there's grouping for the sake of formatting. Nor am I sure if I want a toggle, but it seems the best way to reduce unpredictable behavior with drag-and-drop.

If you see the video you may notice that 'Status' shortened to 1 column, you may want several text fields adjacent but can't unless they're in a container (a group). If you group 3 fields together, then they can be displayed next to the column.

**Comment by Nitwel** [comment]
Hey heres another Idea how it would be possible for more creative aligning of fields:

![live edit](https://user-images.githubusercontent.com/22577866/52902647-066c3400-3214-11e9-87e1-4d62a38419e7.png)

After selecting 1 field:
![live edit 2](https://user-images.githubusercontent.com/22577866/52902658-226fd580-3214-11e9-8143-c126c2b4e8c0.png)

**Comment by ceruulean** [comment]
Some other ideas for a field arranger:
* Widths are saved as a percentage, so the admin can adjust the sizes freehand. Between 0.33 and 0.34 (inclusive) assumed as 1/3 (if using integers, then 33 and 34 are assumed to be 1/3). You can define additional CSS globals for 1/3 and 2/3.

> This makes admin responsible for creating a sensible layout, Directus just provides tools.
* Enable quarter, third, or both guidelines to snap field sizes.

**Comment by Fusseldieb** [comment]
Came here by searching on Google. I read the entire thread, but still don't know exactly why Directus only allow 1/2 or 2/2 controls.

![image](https://user-images.githubusercontent.com/4715129/97743625-b0f9b880-1ac4-11eb-81a0-2a960e702794.png)

Why are the other 2 spaces blank?

---------

Wouldn't it be better if users could just type in the "classes", similar to bootstrap, for each field?
This would allow for incredible flexibility and customization.

Like this:
![image](https://user-images.githubusercontent.com/4715129/97744106-704e6f00-1ac5-11eb-8e52-b86c738bd7f0.png)
(I know it isn't bootstrap, but maybe importing the bootstrap-grid for that would be useful!?) 

--------

But as always, thanks for this incredible piece of open-source software. I'm loving it!

Thanks in advance.

**Comment by rijkvanzanten** [comment]
@Fusseldieb The issue with the spacing on the grid like that is responsiveness, but more importantly: hardly any interface would have a good UX on a size that small. It works for textarea/textinputs, but as soon as you start using any of the more "involved" interfaces like a map, wysiwyg, relational interfaces, translations etc etc etc, it breaks down..

**Comment by Fusseldieb** [comment]
> @Fusseldieb The issue with the spacing on the grid like that is responsiveness, but more importantly: hardly any interface would have a good UX on a size that small. It works for textarea/textinputs, but as soon as you start using any of the more "involved" interfaces like a map, wysiwyg, relational interfaces, translations etc etc etc, it breaks down..
> 
> Re: responsiveness, if we have a 4 (or more, like 8 in your example) grid, we'll have to figure out how to scale it down for non-fullscreen examples, should it start wrapping? Does it go to mobile stack super "early"? etc etc

Thanks for the fast reply!

Well, Bootstap has classes for almost everything to become the size or even location needed. It has offsets, different break-points, etc. Leaving that choice to the user shouldn't be an issue, I think.
Furthermore, since the code can access classes and "read" them too, one could go further and add ""classes"" which indicate different breakpoints for the field (in your example; when to go 'mobile-friendly', etc).

**Comment by rijkvanzanten** [comment]
> Leaving that choice to the user shouldn't be an issue, I think.

In my experience, it kinda is actually! 

That kind of flexibility is great for the minority that really knows how to configure it well. For the others though, it becomes way too easy to royally screw up the UX of the platform. 

It's this kinda thing why we started adhering to [our "80/20 rule"](https://docs.directus.io/getting-started/contributing.html#feature-requests)

**Comment by Fusseldieb** [comment]
> It's this kinda thing why we started adhering to [our "80/20 rule"](https://docs.directus.io/getting-started/contributing.html#feature-requests)

Ohh, didn't see that!

Well, maybe in the future, who knows :)

But it kinda bugs me that only half of the screen is filled. I'm currently doing a frontend with Vue (which has 4 fields, instead of just 2), but it's wayyy to much work, considering that it's already done in Directus...

**Comment by rijkvanzanten** [comment]
Yeah I'm always down to discuss of course, we could come up with some sort of way to allow interfaces to force a range of sizes, and offer a "simple" setup with an advanced overrides mode or whatever. I don't think it's as simple as you might think on first thought in a "has to do everything for everybody everywhere" kinda platform like Directus, but there's a solution to every problem 🙂
